### PR TITLE
Set target fix

### PR DESCRIPTION
**Pull request by willtunnels**
## Description

I added a new struct, `inkwell::targets::TargetTriple` that models a target triple more explicitly in the type system. I fixed the bug where `Module::set_target` and `Module::get_target` were operating on a target rather than a target triple. I also updated all of the functions in the inkwell API which deal with target triples (I think I got all of them at least) to use the `TargetTriple` struct. Corresponding changes were made to the documentation.

`TargetTriple` is a thin wrapper around an `LLVMStringOrRaw`. I modeled `TargetTriple` after `DataLayout`, which achieves a similar aim albeit for a different part of the API. So, even though there are a good number of breaking changes here, the end result is consistent with other parts of the already existing API.

The only part of the implementation I'm not completely satisfied with is the two string copies that have to be made in `TargetTriple::create`, which is the user facing constructor. But, even though this isn't ideal, it should probably be OK because it's unlikely to be on the critical path and, on the off chance it is, things are probably still OK.

## Related Issue

[Fixes this issue.](https://github.com/TheDan64/inkwell/issues/149)

## How This Has Been Tested

There were already test cases for all (most?) of the functions I modified. I updated the existing test suite where needed to be compatible with the new API. I ran `cargo test` and everything passed.

## Breaking Changes

- `Module::set_target` takes a `&TargetTriple` rather than a `&Target`
- `Module::get_target` returns a `TargetTriple` rather than a `Target`
- `Target::create_target_machine` takes a `&TargetTriple` rather than a ` &str` to specify the target triple
- `Target::from_triple` takes a `&TargetTriple` rather than a `&str`
- `TargetMachine::get_triple` returns a `TargetTriple` rather than an `LLVMString`
- `TargetMachine::get_default_triple` returns a `TargetTriple` rather than an `LLVMString`
- `TargetMachine::normalized_target_triple` takes a `&TargetTriple` rather than an `Either<&str, &CStr>` and returns a `TargetTriple` rather than a `LLVMString`

## Checklist

- [x] I have read the [Contributing Guide](https://github.com/TheDan64/inkwell/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: src/targets.rs [hunk 1]
```diff
@@ -96,6 +96,63 @@ impl Default for InitializationConfig {
...
+
+    pub(crate) fn new_borrowed(target_triple: *const ::libc::c_char) -> TargetTriple {
+        debug_assert!(!target_triple.is_null());
+
+        TargetTriple {
+            target_triple: LLVMStringOrRaw::Borrowed(target_triple),
+        }
+    }
+
+    pub fn create(target_triple: &str) -> TargetTriple {
```
Comment (TheDan64) [inline src/targets.rs]: It's not really clear to me what advantage `TargetTriple` has over just passing a string since you can just give `create` any old invalid string? It seems like the same problem as before?

Comment (willtunnels) [reply src/targets.rs]: I was thinking of this as an application of [newtype pattern](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html). The advantage is just that it can help people avoid mixing up strings in their code that represent different things.

One could theoretically add some validation in the future though I'm not actually sure if it's feasible.

Comment (TheDan64) [reply src/targets.rs]: I'm not sure that it is feasible to add validation because you can create custom triples I think? Or at least Rust lets you define triple specifications. Not sure how that works when it gets to LLVM

Comment (TheDan64) [reply src/targets.rs]: I need to look through this again and think on it some more

File: src/targets.rs [hunk 2]
```diff
@@ -96,6 +96,63 @@ impl Default for InitializationConfig {
...
+        let c_string = CString::new(target_triple).expect("Conversion to CString failed unexpectedly");
```
Comment (TheDan64) [inline src/targets.rs]: Nit: This will allocate twice, I think? You should just use `create_from_str` on `target_triple`?

Comment (willtunnels) [reply src/targets.rs]: My thinking was this:
What if the `&str` does not end with a `'\0'`? In general I would not expect it to. So, I would need to append `'\0'` to the input. To append `'\0'` I would need an owned copy, which would reallocate anyway. Am I missing something here?

Comment (TheDan64) [reply src/targets.rs]: You're right, but there's gotta be a better way..

File: src/targets.rs (hunk 1 above)
Comment (TheDan64) [reply src/targets.rs]: I think there's value in a newtype when you can guarantee that only valid `TargetTriple` are generated, which we cannot

File: src/module.rs [hunk 3]
```diff
@@ -321,59 +321,56 @@ impl<'ctx> Module<'ctx> {
...
     /// ```
-    pub fn get_target(&self) -> Option<Target> {
+    pub fn get_target(&self) -> TargetTriple {
         // REVIEW: This isn't an owned LLVMString, is it? If so, need to deallocate.
         let target_str = unsafe {
             LLVMGetTarget(self.module.get())
         };
 
-        Target::from_name_raw(target_str)
+        TargetTriple::new_borrowed(target_str)
```
Comment (TheDan64) [inline src/module.rs]: A big problem is that we're leaking a raw pointer into the wild. Really this should be `-> &TargetTriple` for borrowed triples

File: src/targets.rs (hunk 1 above)
Comment (willtunnels) [reply src/targets.rs]: I see both perspectives here, so I'm not super inclined to argue hard for the newtype. It's my personal inclination, but you've got a point. Just let me know what you decide on this so I can finish up the PR and we can merge it.

File: src/module.rs [hunk 4]
```diff
@@ -321,59 +321,56 @@ impl<'ctx> Module<'ctx> {
...
+    /// assert_eq!(module.get_target(), TargetTriple::create(""));
     ///
-    /// module.set_target(&target);
+    /// module.set_target(&triple);
     ///
-    /// assert_eq!(module.get_target().unwrap(), target);
+    /// assert_eq!(module.get_target(), triple);
     /// ```
-    pub fn set_target(&self, target: &Target) {
+    pub fn set_target(&self, target_triple: &TargetTriple) {
```
Comment (TheDan64) [inline src/module.rs]: We should change these to `get/set_target_triple` for clarity, despite LLVM's poorly named function that implies an actual Triple object

File: tests/all/test_module.rs [hunk 5]
```diff
@@ -316,27 +316,25 @@ fn test_print_to_file() {
 
 #[test]
 fn test_get_set_target() {
-    Target::initialize_x86(&Default::default());
```
Comment (TheDan64) [inline tests/all/test_module.rs]: Why was this removed? This is generally to ensure LLVM's globals are initialized properly

File: src/targets.rs (hunk 2 above)
Comment (willtunnels) [reply src/targets.rs]: This is abstraction breaking so probably best to avoid, but if you look at the LLVM source code you can see that creating a message just does a `strdup` and freeing it just does a `free` so you could actually pretended any old owned cstring was allocated by LLVM.

File: tests/all/test_module.rs (hunk 5 above)
Comment (willtunnels) [reply tests/all/test_module.rs]: The test runs fine without it. Getting/setting the module's target triple does not touch the `TargetRegistry`. It's probably better to put it back in though, as a defensive measure, because I'm not sure that no code path touches the `TargetRegistry` given a module on which you've set the target triple.

Comment (willtunnels) [reply tests/all/test_module.rs]: LLVM's sparse documentation makes it hard for me to tell exactly what preconditions the APIs have

File: src/module.rs (hunk 4 above)
Comment (willtunnels) [reply src/module.rs]: Apparently that's the name the corresponding function in the [C++ API](https://llvm.org/doxygen/classllvm_1_1Module.html#af66bad1510da8a71d20002609d3b4a25) has

File: src/module.rs (hunk 3 above)
Comment (willtunnels) [reply src/module.rs]: I think a similar problem may exist with DataLayout since the origin of the problem is really with LLVMStringOrRaw not having a proper lifetime on its borrow variant. Maybe the solution is to define an LLVMStr that implements ToOwned and stick it inside an actual Cow?

Comment (willtunnels) [reply src/module.rs]: I fixed by making a copy. I can't imagine that this is a hot path anyway and doing something more efficient really increased code complexity. Similarly, there are more efficiently ways to implement `TargetTriple;:create` (maybe you could hold an enum of `LLVMString` and `CString`) but I'm not really sure they're worth it. Also, as mentioned above, there is definitely a leakage issue with `DataLayout::get_borrowed_data_layout`. I guess I will make a separate issue for this...

File: src/module.rs (hunk 4 above)
Comment (willtunnels) [reply src/module.rs]: I called it `get/set_triple` for consistency with the rest of the `inkwell` api.

File: src/module.rs (hunk 3 above)
Comment (TheDan64) [reply src/module.rs]: Sounds good. We can always optimize the copy out in the future. Please just add a TODO comment so that we don't forget.

Comment (TheDan64) [reply src/module.rs]: And maybe an issue too

**Comment by TheDan64** [comment]
It looks like something needs to be fixed for most of the older LLVM versions

**Comment by willtunnels** [comment]
I have been extremely busy with school. Sorry for the delay. I went back and fixed the issue with old LLVM versions.

**Comment by TheDan64** [comment]
No problem. And thanks, I'll take a look soon. Sorry about the whack-a-mole situation. I usually just rely on the travis tests to tell me which imports aren't correct and then just look up their paths in llvm-sys.

File: src/targets.rs [hunk 6]
```diff
@@ -96,6 +95,53 @@ impl Default for InitializationConfig {
...
+        TargetTriple {
+            triple,
+        }
+    }
+
+    pub fn create(triple: &str) -> TargetTriple {
+        let c_string = CString::new(triple).expect("Conversion to CString failed unexpectedly");
+
+        TargetTriple {
+            triple: LLVMString::create_from_c_str(c_string.as_c_str())
```
Comment (TheDan64) [inline src/targets.rs]: You should just do `LLVMString::create_from_str(triple)` so as to not allocate with `CString`

Comment (willtunnels) [reply src/targets.rs]: Hmm, when tests started failing I remembered the reason why I did this: if the user's &str is not NUL terminated then calling LLVMString::create_from_str will fail (and in general we don't expect it to be). This is a pretty fundamental Rust and C interop problem and I'm curious about your thoughts on it. We could avoid one of the allocations by cheating because we know the implementation of `LLVMCreateMessage` and `LLVMDisposeMessage`...

Comment (TheDan64) [reply src/targets.rs]: Ack, you're right. I keep forgetting about this. lol. Let's leave it as it is then and I'll merge
